python/pipeline/utils/clocktools.py

```diff
@@ -0,0 +1,996 @@
+                  
+                * Recording points where the time value is NaN are dropped from analysis/processing
+    """
+
+    ##
+    ## Fetch source and target times, along with converting between Stimulus or Behavior clock if needed
+    ##
+
+    source_times, target_times = fetch_timing_data(
+        scan_key, source_type, target_type, debug
+    )
+
+    ##
+    ## Convert indices to a list of numbers if argument equals None or a Boolean mask
+    ##
+
+    if indices is None:
+        indices = np.arange(len(source_times))
+    elif type(indices[0]) == bool:
+        indices = np.where(indices)[0]
+    else:
+        ## Check for duplicates if manually entered
+        if len(np.unique(indices)) != len(indices):
+            msg = (
+                f"Duplicate entries found for provided indice array! "
+                f"Try to fix the error or use np.unique() on indices array."
+            )
+            raise PipelineException(msg)
+
+    ##
+    ## Convert source indices to time boundaries, then convert time boundaries into target indices
+    ##
+
+    ## Convert indices into start/end times for each continuous fragment (incrementing by 1)
+    time_boundaries = find_time_boundaries(indices, source_times, drop_single_idx)
+    target_indices = []
+    single_idx_count = 0
+
+    ## Loop through start & end times and create list of indices corresponding to that block of time
+    for [start, end] in time_boundaries:
+        target_idx = np.where(
+            np.logical_and(target_times >= start, target_times <= end)
+        )[0]
+        if len(target_idx) < 2:
+            if drop_single_idx:
+                single_idx_count += 1
+            else:
+                msg = (
+                    f"Event of length {len(target_idx)} found. "
+                    f"Set drop_single_idx to True to suppress these errors."
+                )
+                raise PipelineException(msg)
+        else:
+            target_indices.append(target_idx)
+
+    if debug:
+        print(f"Indices converted. {single_idx_count} events of length 0 or 1 dropped.")
+
+    ##
+    ## Interpolate related signal if requested, else just return the target_indices found.
+    ##
+
+    if return_interpolate:
+
+        ## Create full interpolated signal
+        interpolated_signal = interpolate_signal_data(
+            scan_key, source_type, target_type, source_times, target_times, debug=debug
+        )
+
+        ## Split indices given into fragments based on which ones are continuous (incrementing by 1)
+        source_signal_fragments = []
+        for idx_fragment in target_indices:
+            source_signal_fragments.append(interpolated_signal[idx_fragment])
+
+        ## If full signal is converted, remove wrapping list
+        if len(source_signal_fragments) == 1:
+            source_signal_fragments = source_signal_fragments[0]
+
+        return source_signal_fragments
+
+    else:
+
+        return target_indices
+    
+
```

Comment:
All of the above comments also refer to the rest of the convert_clocks_*_to_*, since they all follow pretty much the same documentation/outline.